x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java

```diff
@@ -352,74 +357,71 @@ public String toString() {
+                    final IndexInput input = openInput(file.physicalName(), CachedBlobContainerIndexInput.CACHE_WARMING_CONTEXT);
                     assert input instanceof CachedBlobContainerIndexInput : "expected cached index input but got " + input.getClass();
 
-                    final long numberOfParts = cacheFile.numberOfParts();
-                    final CountDown countDown = new CountDown(Math.toIntExact(numberOfParts));
-                    for (long p = 0; p < numberOfParts; p++) {
-                        final int part = Math.toIntExact(p);
-                        // TODO use multiple workers to warm each part instead of filling the thread pool
-                        executor.execute(new AbstractRunnable() {
-                            @Override
-                            protected void doRun() throws Exception {
-                                ensureOpen();
-
-                                logger.trace("warming cache for [{}] part [{}/{}]", fileName, part, numberOfParts);
-                                final long startTimeInNanos = statsCurrentTimeNanosSupplier.getAsLong();
-
-                                final CachedBlobContainerIndexInput cachedIndexInput = (CachedBlobContainerIndexInput) input.clone();
-                                cachedIndexInput.prefetchPart(part); // TODO does not include any rate limitation
-
-                                logger.trace(
-                                    () -> new ParameterizedMessage(
-                                        "part [{}/{}] of [{}] warmed in [{}] ms",
-                                        part,
-                                        numberOfParts,
-                                        fileName,
-                                        TimeValue.timeValueNanos(statsCurrentTimeNanosSupplier.getAsLong() - startTimeInNanos).millis()
-                                    )
-                                );
-                            }
-
-                            @Override
-                            public void onFailure(Exception e) {
-                                logger.trace(
-                                    () -> new ParameterizedMessage(
-                                        "failed to warm cache for [{}] part [{}/{}]",
-                                        fileName,
-                                        part,
-                                        numberOfParts
-                                    ),
-                                    e
-                                );
-                            }
-
-                            @Override
-                            public void onAfter() {
-                                if (countDown.countDown()) {
-                                    IOUtils.closeWhileHandlingException(input);
-                                }
-                            }
-                        });
+                    final int numberOfParts = Math.toIntExact(file.numberOfParts());
```

Comment:
I'm wondering if we should just special case `numberOfParts == 1` here? It's the by far most common case and we could simply run it straight on the current thread instead of creating another task. Currently, we're creating twice as many tasks on the pool as necessary with this change.
No need for a lot of code here even, could just keep it short and use the direct executor service if `numberOfParts == 1` and be happy with it?

Reply:
Makes perfect sense, I pushed 47b2f0e5e175111adc1937d0884b2246579f3727